### ISSUE TITLE
Windows builds

### ISSUE THREAD

**Issue by falkTX** [jackaudio/jack2#287]
as requested, the next version was bumped to 1.9.12
I would like to figure out windows builds for this release

@kmatheussen do you think you can provide those.
If yes, let me know if any changes are needed. After you get it to build, I'll tag it so you can upload the binaries.
Thank you in advance.

**Comment by kmatheussen** [comment]
Hi, there is an issue with creating the installer. If I remember correctly, jack for windows currently uses a ready-made program to create the installer which probably can't be automated when cross-compiling. I think we need a better solution to create an installer.

Bit there is also an alternative, and more radical, solution. Namely to not use an installer at all, and not having to install jack at all. We could require programs linking to jack on windows to include the dll-files, plus require windows programs to weak-link jack. The idea is that windows programs links to a small static library that obtains the paths to the used libraries of the currently running jack. And if jack is not running, the program itself will provide the paths to the used libraries of the currently running jack.

This solution might even require less work than to fix the installer. Another advantage is that people don't have to install jack anymore. Older programs still need to have a jack installed though, but I think it's possible to do this without breaking older programs.

**Comment by falkTX** [comment]
I don't agree with your proposed solution, if I understood it properly that is..
You want to ship jack with each application?

We don't have to automate everything.
I will try to do mingw builds myself too, but I need to update it first because my version is too old now.

**Comment by kmatheussen** [comment]
> I don't agree with your proposed solution, if I understood it properly that is..
> You want to ship jack with each application?

Yes, what's wrong with that?

> We don't have to automate everything.

Well, I really don't like the work flow of clicking some buttons and then hopefully generate the correct thing. It's likely that you will do something wrong some day. It's better to have a script.

**Comment by kmatheussen** [comment]
> Yes, what's wrong with that?

But this can be made optional. Only if you need to start the server, you must include jack.

**Comment by y-fujii** [comment]
Just for a reference, I made a MinGW package build script.  See <https://github.com/y-fujii/MINGW-packages/tree/jack2/mingw-w64-jack2-git>.  I plan to open a pull request to MinGW/MSYS2 official repository later.  Any suggestions are welcome.  It is not heavily tested (although I am interested in JACK on Windows, my main environment is Linux...), but confirmed to work with Carla (many thanks to @falkTX) and some my personal programs.

This build script is suitable for MSYS2 environment, but not for an official JACK2 Windows installer.  I think it needs some works if it is used for official installer.

- libjack(64).dll should be statically-linked.
	- libjack.dll in MinGW package depends on many shared objects such as libstdc++.dll.  It is preferred for MSYS2 environment, but not for an official installer because official JACK installer places libjack.dll to c:/windows.

- libjack.lib
	- Current libjack.lib in the installer is generated from libjack.def, which has ordinal numbers "@0, @1, ...".  It causes a DLL link problem because functions are looked up by ordinal numbers, not by symbol names.  I think this problem is easy to solve:  Remove "@0, @1, ..." from libjack.def and regenerate libjack.lib by dlltool.

	      dlltool -d libjack.def -l libjack.lib

- JackRouter
	- I confirmed that JackRouter.dll can be compiled with MinGW64 without massive patches, but not tested yet.

**Comment by Schroedingers-Cat** [comment]
I fixed some issues with the installer and also updated the installer project files to the latest version of the free CreateInstaller software. This allows for silent switches and a completely automated setup (and even automatic updates if used in conjunction with chocolatey).   
However, for testing I need all compiled Jack files which seems nightmarish to get done on Win10 with VS2017 as you either need a very special version of the MinGW compiler or the VC6 compiler and IDE.  

https://github.com/jackaudio/jack2/pull/318 looks really promising in this regard but seems to need a rebase on the current master. The proposed commits from the PR look very clear to me (though I don't have experience with this repo's code so I could be missing something) and I see no reason to that rebase can't be done by a maintainer.  

Anyway, if somebody could upload the compiled files in the original folder structure, I could continue to test the updated installer files, make a PR for this and create a chocolatey package out of the silent switches from the installer.

**Comment by clvn** [comment]
These are the major classes of problems encountered when attempting to build with MSVC:

## Compiler errors

- Differences in platform headers (stdint, unistd, alloca, usleep, strcasecmp, getopt...)
- Missing namespace usings

## Dependencies

- Cannot rely on the OS' package manager to get portaudio, libsamplerate, tre, pthread
- The closest to a package manager on Windows is vcpkg, which can be automated to build all dependencies required by Jack

## Shared library symbols import/export (the big one)

- GCC/MingW marks all symbols "visible", shared libraries are linked at run time
- MSVC requires explicitly marking public symbols as either being exported or imported from a DLL, shared libraries are linked at compile time

## Build system

- Need to adapt compiler flags required by MSVC
- Need to update WAF's resource compiler to embed version/author metadata into DLLs (i.e *.rc)

---

Both mine and justnope's branches fix all of these. Rather than attempting to merge everything at once, suggest to split into well-explained, bite-sized PRs to increase the chances of a quick merge.  Once the repo builds with MSVC, we can look at creating a Windows installer via CI as well.

---

For example, the reasoning behind this commit https://github.com/clvn/jack2/commit/e7b06bf685676db25c11e07367fc9c5043a10a68 :

This fixes compiler errors when building a jack client on Windows with Visual Studio 2010 or newer, and stdint.h/cstdint is included before jack.h. F.ex the following code:

```c++
#include <sstream>
#include <jack/jack.h>
```

Errors with the following:

```
1>d:\code\vcpkg\installed\x86-windows\include\jack\systemdeps.h(73): error C2371: 'int8_t': redefinition; different basic types
1>c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.16.27023\include\stdint.h(17): note: see declaration of 'int8_t'
1>d:\code\vcpkg\installed\x86-windows\include\jack\systemdeps.h(77): error C2371: 'int32_t': redefinition; different basic types
1>c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.16.27023\include\stdint.h(19): note: see declaration of 'int32_t'
1>d:\code\vcpkg\installed\x86-windows\include\jack\systemdeps.h(78): error C2371: 'uint32_t': redefinition; different basic types
1>c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.16.27023\include\stdint.h(23): note: see declaration of 'uint32_t'
```

The _MSC_VER macro has a value of 1600 in Visual Studio 2010, so the commit changes to use stdint.h when it should be present.

References:

* https://docs.microsoft.com/en-us/previous-versions/mt743084(v=vs.140)#errors-involving-stdinth-integral-types
* https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=vs-2017

**Comment by davidebeatrici** [comment]
Types conflict reported in #353.

**Comment by gvanem** [comment]
@clvn Shared library symbols import/export (the big one)

Yes, that's the big showstopper for me too. Trying to build using MSVC. or my favourite choice , `clang-cl` turns out to be a nightmare due to the missing `__declspec(dllexport)` and `__declspec(dllimport)`. 

These `SERVER_EXPORT` and `LIB_EXPORT` mean next to nothing for a new-comer like me.

I tried to extract the details from the `windows/*.cbp` files to get a sense of the build-process. 
It seems to me several .cpp files are used in at least 2 .DLLs. Why?

IMHO something like this (in `common/export.h`) would be clearer:
```c
#if defined(_WIN32) && !defined(JACK_STATIC)
  #define DLL_EXPORT __declspec(dllexport)
  #define DLL_IMPORT __declspec(dllimport)

  #ifdef BUILDING_JACK_DLL
    #define JACK_API DLL_EXPORT
  #else 
    #define JACK_API DLL_IMPORT  
  #endif
#else  
  #define JACK_API
#endif

/* Similar for the other .DLLs */
```

**Comment by coffeeandspaceships** [comment]
I've just discovered develop branch and master branch are different for at least one case within "Differences in platform headers." On [master](https://github.com/jackaudio/jack2/blob/master/windows/JackSystemDeps_os.h) there is an undefended UINT32_MAX which cause a redefinition error on my laptop. This has been addressed on [develop](https://github.com/jackaudio/jack2/blob/develop/windows/JackSystemDeps_os.h)

This will interfere with creating pull requests to do the compiler error cleanups - which one should we use as the base for this purpose? I guess since @kmatheussen is actively working on Windows issues on the develop branch we should stay in sync with that?

**Comment by falkTX** [comment]
Always use the develop branch for new stuff
